### ISSUE TITLE
[RFC]: KL-divergence evaluation tool

### ISSUE THREAD

**Issue by mratsim** [vllm-project/llm-compressor#2031]
I suspect some of the quant I create are suboptimal, it would be extremely helpful to have a way to evaluate how well the quantized model reproduce the original's raw probability distribution before any sampler.

Hence it would be helpful to have a tool that reports KL-divergence with base model, KLD(A, B) and KLD(B, A) since KL divergence is asymmetric.

**Comment by kylesayrs** [comment]
Hi @mratsim!

FYI the GPTQModifier prints weight MSE loss with the base model for GPTQ. It is somewhat known that down_proj layers are difficult to quantize, and there are some internal efforts at Red Hat to develop a "sensitivity analysis tool" which measures the kurtosis of given weights to determine which are most difficult to quantize.

Checking MSE/KL loss between base model weights and quantize model weights should be a fairly straight forward script.

**Comment by czhang165** [comment]
I'd like to claim this issue! KL divergence is probably the most beautiful function in the history of deep learning.

Primary metric: Output KL-divergence
Secondary: Per-layer activation MSE for debugging which layers cause the most drift
The idea is that we want to see where the error propagation starts to emerge, which is usually not reflected in the weights value.
Is there an existing evaluation framework/class I can hook into?

**Comment by kylesayrs** [comment]
Hi all,

The RedHat research team is currently working on a tool to perform quantization analysis using metrics such as MSE and KL-Divergence, currently being explored by @krishnateja95.

@czhang165 I'll post more information on this ticket soon!

**Comment by NJX-njx** [comment]
Hi, I'd like to work on this. I plan to implement a KL-divergence evaluation tool that:
1. Loads both the base model and quantized model
2. Runs forward passes on a calibration/eval dataset to collect logits
3. Computes KLD(base, quant) and KLD(quant, base) since KL divergence is asymmetric
4. Reports per-layer and aggregate statistics

Will integrate it as a standalone utility under the llmcompressor tools/evaluation module. Happy to discuss the API design before implementing.

**Comment by phaelon74** [comment]
@kylesayrs @czhang165 @NJX-njx 

See my RFC here: https://github.com/vllm-project/vllm/issues/35962
See my PR for VLLM: https://github.com/vllm-project/vllm/pull/35961

My PR works in VLLM, Captures ON GPU versus back and forth, and then stores KLD reference local for subsequent runs.  On an RTX PRO 6000 Blackwell Workstation card, it takes ~2-3 minutes to complete.

If built as a separate Application and/or build this into LLM_Compressor, we should follow Turbo's lead from EXL3, which is what my implementation is based off of.  2048 Context window with a Stride of 512, and EVERY POSITION, in EVERY WINDOW is analyzed.

Aes will be PRing llama.cpp shortly to do the same thing, with the same alignment, so we will be as close to apples to apples to apples when it comes to EXL3 to GGUF, to GPTQ to AWQ to NVFP4, etc.

Also, KLD shows without question, that NVFP4 under 400B, needs QAD to come anywhere close to Nvidia's claims of ~1% of FP8.  Doing QAD on Llama3.1-8B-Instruct now and have ~5 days left.  Will then add it to the graph.  I can tell you though, it's going to be WAY better.  440,000 tokens on QAD and KLD went down by ~20% on it, so QAD is a MUST for models under 400B when NVFP4 is the quant scheme.  See below picture:

Also, look at the Scores below.  Those sharper the KLD, that's me adding better Datasets to my AWQ quanting.

Also, KLD shows that quantizing Activations does degrade a model.  W8A8 (FP8_BLOCK) is worse than W8A16, which is to be expected.

<img width="1500" height="900" alt="Image" src="https://github.com/user-attachments/assets/f27a2b42-d226-453b-ad7e-7cbac92bdce7" />

Command I use to run it:
`python3 examples/score_mode_kld.py \
    --model /media/fmodels/TheHouseOfTheDude/Llama-3.1-8B-Instruct_Compressed-Tensors/FP8_INT4 \
    --reference-model /media/fmodels/meta-llama/Llama-3.1-8B-Instruct/ \
    --dataset wikitext \
    --dataset-config wikitext-2-raw-v1 \
    --context-length 2048 \
    --stride 512 \
    --tensor-parallel-size 2 \
    --gpu-memory-utilization 0.30`

KLD RESULTS:
FP32 : 0.0 (30G)

FP8-INT4 (6.2G)
Results:
  Mean KLD: 0.033707
  Total positions: 204700
  Time elapsed: 38.05 seconds
  Positions/second: 5380.21
  
W4A16_GS128 (5.4G)
Results:
  Mean KLD: 0.076226
  Total positions: 204700
  Time elapsed: 39.29 seconds
  Positions/second: 5210.26

W4A16_GS128 (5.4G) {DS02 - /media/fmodels/TheHouseOfTheDude/Llama-3.1-8B-Instruct/W4A16/}
Results:
  Mean KLD: 0.076194
  Total positions: 204700
  Time elapsed: 962.45 seconds
  Positions/second: 212.69
  
W4A16_GS128 (5.4G) {DS02 - /media/fmodels/TheHouseOfTheDude/Llama-3.1-8B-Instruct_CompressedTensors/W4A16/}
Results:
  Mean KLD: 0.072525
  Total positions: 204700
  Time elapsed: 137.45 seconds
  Positions/second: 1489.30
(venv) phaedawg@d011sd02:~/nightly-kld/vllm$

Llama3.1-8B-Instruct - Quantization Scheme W4A16_GS32 (Size on Disk 5.7G)
Run on DS01 (Two RTX Pro 6000 Workstation Blackwell)
Results:
  Mean KLD: 0.048686
  Total positions: 204700
  Time elapsed: 39.16 seconds
  Positions/second: 5227.31

Run on DS02 (Four RTX 3090) - New code after refactor
Results:
  Mean KLD: 0.048687
  Total positions: 204700
  Time elapsed: 139.13 seconds
  Positions/second: 1471.26

NVFP4 {nvidia/Llama-3.1-8B-Instruct-NVFP4}
Results:
  Mean KLD: 0.101230
  Total positions: 204700
  Time elapsed: 2333.90 seconds
  Positions/second: 87.71

NVFP4 (5.7G)
Results:
  Mean KLD: 0.109275
  Total positions: 204700
  Time elapsed: 35.43 seconds
  Positions/second: 5778.28

NVFP4_New (5.7G)
Results:
  Mean KLD: 0.089775
  Total positions: 204700
  Time elapsed: 35.88 seconds
  Positions/second: 5705.64
  
NVFP4-QAD {Partial, only 440,000 tokens.  Needs ~500,000,000 to 2,500,000,000 for true alignment) (5.7G)
Results:
  Mean KLD: 0.084104
  Total positions: 204700
  Time elapsed: 331.51 seconds
  Positions/second: 617.47

  
W8A16_GS128 (8.6G)
Results:
  Mean KLD: 0.000899
  Total positions: 204700
  Time elapsed: 53.79 seconds
  Positions/second: 3805.66

W8A16_GS32 (8.9G)
Results:
  Mean KLD: 0.000813
  Total positions: 204700
  Time elapsed: 40.88 seconds
  Positions/second: 5006.79


W8A8_FP8_BLOCK (8.5G)
Results:
  Mean KLD: 0.006547
  Total positions: 204700
  Time elapsed: 43.45 seconds
  Positions/second: 4710.75

**Comment by HDCharles** [comment]
Hey this looks cool, @phaelon74 reach out to me in the vllm slack and we can discuss this. 

In the meantime a few thoughts. 

1) I think it could be useful to have a more streamlined method of evaluation.
2) @phaelon74 the vllm runtime team is the one who'll review your issue and PR but i don't think its likely to land to be honest, its a lot of code in hot paths though you can see what they say.
3) I think it might be feasible to A) run both models in vllm as is, B) record the penultimate layer of activations C) load the output layers of both models to obtain logits and then calculate KL divergence like that.
